### PR TITLE
sds: Factor out new function getServiceIdentitiesFromCert()

### PR DESCRIPTION
**Pull request by draychev**
This PR factors out a new function called `getServiceIdentitiesFromCert()`.

This, I believe shows clearly the purpose of it - to compute the Subject Alt Names for the `MatchSubjectAltNames` Envoy config.

I think the purpose of this block of code, prior to this PR, is lost in the sheer volume of code.

---

In this PR:
 - carved out a block of code into `getServiceIdentitiesFromCert()`
 - no functional code changes
 - removed `switch / case` -- using `if / return` instead - for simplicity (fewer indentations)
 - renamed `errGotUnexpectedCertRequest` to `errCertMismatch` for clarity
 - renamed `getSubjectAltNamesFromSvcAccount` to `getSubjectAltNamesFromSvcIdentities` for correctness

### CODE REVIEW

File: pkg/envoy/sds/response.go [hunk 1]
```diff
@@ -145,58 +144,66 @@ func (s *sdsImpl) getRootCert(cert certificate.Certificater, sdscert envoy.SDSCe
...
+	svcIdentities, err := getServiceIdentitiesFromCert(sdsCert, s.serviceIdentity, s.meshCatalog)
+	if err != nil {
+		return nil, err
+	}
+
+	secret.GetValidationContext().MatchSubjectAltNames = getSubjectAltNamesFromSvcIdentities(svcIdentities)
+	return secret, nil
+}
+
+func getServiceIdentitiesFromCert(sdsCert envoy.SDSCert, svcIdentity identity.ServiceIdentity, mc catalog.MeshCataloger) ([]identity.ServiceIdentity, error) {
```
Comment (shashankram) [inline pkg/envoy/sds/response.go:156]:
```suggestion
func getServiceIdentitiesFromCert(sdsCert envoy.SDSCert, svcIdentity identity.ServiceIdentity, cataloger catalog.MeshCataloger) ([]identity.ServiceIdentity, error) {
```

File: pkg/envoy/sds/response.go [hunk 2]
```diff
@@ -145,58 +144,66 @@ func (s *sdsImpl) getRootCert(cert certificate.Certificater, sdscert envoy.SDSCe
...
 		if err != nil {
 			log.Error().Err(err).Msgf("Error listing service accounts for service %s", meshSvc)
 			return nil, err
 		}
-		secret.GetValidationContext().MatchSubjectAltNames = getSubjectAltNamesFromSvcAccount(svcIdentities)
+		return svcIdentities, nil
+	}
 
-	case envoy.RootCertTypeForMTLSInbound:
+	if sdsCert.CertType == envoy.RootCertTypeForMTLSInbound {
```
Comment (shashankram) [inline pkg/envoy/sds/response.go:175]: Instead of a bunch of `if` clauses with returns, a `switch-case` cluase reads better to me. Since the purpose of this PR is to carve out a new function `getServiceIdentitiesFromCert()`, I would avoid refactoring other parts of the code here since it only makes it harder to review.

File: pkg/envoy/sds/response.go (hunk 1 above)
Comment (shashankram) [inline pkg/envoy/sds/response.go:156]: Do you think an extra unit test for this function makes sense? I know it implicitly gets tested when other functions that call it are tested.

Comment (shashankram) [inline pkg/envoy/sds/response.go:156]: Does this function return the identity as an FQDN or the ServiceAccount encapsulated as a ServiceIdentity type?
A comment will help here with an example of what the return value looks like, otherwise, the code is pretty hard to follow.

File: pkg/envoy/sds/response.go [hunk 3]
```diff
@@ -145,58 +144,66 @@ func (s *sdsImpl) getRootCert(cert certificate.Certificater, sdscert envoy.SDSCe
 		return secret, nil
 	}
 
-	// Program SAN matching based on SMI TrafficTarget policies
-	switch sdscert.CertType {
-	case envoy.RootCertTypeForMTLSOutbound:
+	svcIdentities, err := getServiceIdentitiesFromCert(sdsCert, s.serviceIdentity, s.meshCatalog)
```
Comment (shashankram) [inline pkg/envoy/sds/response.go:147]: This function looks confusing because it passes in the proxy's own ServiceIdentity and returns a list of ServiceIdentity. Renaming the variable might clear the confusion.

```suggestion
	svcIdentitiesInCertRequest, err := getServiceIdentitiesFromCert(sdsCert, s.serviceIdentity, s.meshCatalog)
```

File: pkg/envoy/sds/response.go (hunk 2 above)
Comment (draychev) [reply pkg/envoy/sds/response.go:175]: Sure, if you prefer `switch`, I'll keep `switch`.

File: pkg/envoy/sds/response.go (hunk 1 above)
Comment (draychev) [reply pkg/envoy/sds/response.go]: Not yet. This is an intermediary step to further simplification.  I would not test this just yet.

Comment (draychev) [reply pkg/envoy/sds/response.go]: Agreed the area of the code is hard to follow. That's why this needs refactoring.  This is step 1.  We'll continue to plow through this.  I have a few other ideas to offer.

Added a comment for the short time while this function is in existence.

**Review by shashankram — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]
